File: videoprokat/BluRayCarrier.cpp
#include "BluRayCarrier.hpp"
#include <iostream>

BluRayCarrier::BluRayCarrier()
    : VideoCarrier(),
      m_is4K(false),
      m_hasHDR(false),
      m_storageGB(25)
{
}

BluRayCarrier::BluRayCarrier(int inventoryNumber,
                             const std::string& title,
                             const std::string& genre,
                             double rentalPricePerDay,
                             double fullPrice,
                             bool is4K,
                             bool hasHDR)
    : VideoCarrier(inventoryNumber, title, "BluRay", genre, rentalPricePerDay, fullPrice),
      m_is4K(is4K),
      m_hasHDR(hasHDR),
      m_storageGB(is4K ? 50 : 25)
{
    std::cout << "Создан BluRay носитель: " << title 
              << (is4K ? " (4K)" : "") << (hasHDR ? " (HDR)" : "") << "\n";
}

BluRayCarrier::~BluRayCarrier()
{
    std::cout << "Деструктор BluRayCarrier для " << m_title << "\n";
}

void BluRayCarrier::markAsRented()
{
    m_status = "rented";
    std::cout << "BluRay диск " << (m_is4K ? "4K " : "") 
              << "арендован: " << m_title << "\n";
}

void BluRayCarrier::markAsAvailable()
{
    std::cout << "BluRay возвращен: " << m_title << "\n";
    VideoCarrier::markAsAvailable();
}

double BluRayCarrier::calculatePremiumPrice() const
{
    double basePrice = m_rentalPricePerDay;
    if (m_is4K) basePrice *= 1.5;
    if (m_hasHDR) basePrice *= 1.2;
    return basePrice;
}

bool BluRayCarrier::isIs4K() const
{
    return m_is4K;
}

bool BluRayCarrier::hasHDR() const
{
    return m_hasHDR;
}

int BluRayCarrier::getStorageGB() const
{
    return m_storageGB;
}

File: BluRayCarrier.java
package videoprokat;

/**
 * BluRay-носитель - производный класс от VideoCarrier.
 * Демонстрирует наследование и использование виртуальных методов.
 */
public class BluRayCarrier extends VideoCarrier {
    
    private boolean is4K;
    private boolean hasHDR;
    private int storageGB;
    
    public BluRayCarrier() {
        super();
        this.is4K = false;
        this.hasHDR = false;
        this.storageGB = 25;
    }
    
    /**
     * Конструктор с параметрами, вызывает конструктор базового класса.
     */
    public BluRayCarrier(int inventoryNumber, String title, String genre,
                        double rentalPricePerDay, double fullPrice, boolean is4K, boolean hasHDR) {
        super(inventoryNumber, title, "BluRay", genre, rentalPricePerDay, fullPrice);
        this.is4K = is4K;
        this.hasHDR = hasHDR;
        this.storageGB = is4K ? 50 : 25;
    }
    
    /**
     * Перегрузка метода базового класса БЕЗ ВЫЗОВА базового метода.
     * Полностью переопределяет логику.
     */
    @Override
    public void markAsRented() {
        status = "rented";
        System.out.println("BluRay диск " + (is4K ? "4K " : "") + "арендован: " + title);
    }
    
    /**
     * Перегрузка с вызовом базового метода.
     */
    @Override
    public void markAsAvailable() {
        System.out.println("BluRay возвращен: " + title);
        status = "available";
    }
    
    /**
     * Использует protected поля базового класса.
     */
    protected double calculatePremiumPrice() {
        double basePrice = rentalPricePerDay;
        if (is4K) basePrice *= 1.5;
        if (hasHDR) basePrice *= 1.2;
        return basePrice;
    }
    
    @Override
    public int getInventoryNumber() {
        return inventoryNumber;
    }
    
    @Override
    public String getTitle() {
        return title;
    }
    
    @Override
    public String getCarrierType() {
        return carrierType;
    }
    
    @Override
    public String getGenre() {
        return genre;
    }
    
    @Override
    public double getRentalPricePerDay() {
        return calculatePremiumPrice();
    }
    
    @Override
    public double getFullPrice() {
        return fullPrice;
    }
    
    @Override
    public String getStatus() {
        return status;
    }
    
    @Override
    public boolean isAvailable() {
        return "available".equals(status);
    }
    
    public boolean isIs4K() {
        return is4K;
    }
    
    public boolean isHasHDR() {
        return hasHDR;
    }
    
    public int getStorageGB() {
        return storageGB;
    }
}
